File: C/project-euler-2.c
#include <stdio.h>

int SumEvenFibonacciBruteForce(int);
int SumEvenFibonacci(int);

int main(){

  printf("%d\n", SumEvenFibonacciBruteForce(4000000));
  printf("%d\n", SumEvenFibonacci(4000000));

  return 0;
}

// Brute Force Method
// Get all Fibonacci numbers and add to sum if the nth term is even
int SumEvenFibonacciBruteForce(int limit){
  int sum = 0;
  int first_term = 1;
  int second_term = 1;
  int nth_term = 0;

  while(nth_term < limit){
    nth_term = first_term + second_term;
    first_term = second_term;
    second_term = nth_term;

    if(nth_term % 2 == 0)
      sum = sum + nth_term;
  }

  return sum;
}

// Removes checking of nth term if even
// Took advantage of the fact that even numbers
// appear every multiple of 3 term
int SumEvenFibonacci(int limit){
  int sum = 0;
  int first_term = 1;
  int second_term = 1;
  int nth_term = 0;

  while(nth_term < limit){
    nth_term = first_term + second_term;
    first_term = second_term + nth_term;
    second_term = nth_term + first_term;
    
    if(nth_term < limit)
      sum = sum + nth_term;
  }
  return sum;
}

File: python/project-euler-2.py
import time

def solution_one(limit):
  result = nth_term = 0
  first_term = second_term = 1

  while nth_term < limit:
    nth_term = first_term + second_term
    first_term = second_term
    second_term = nth_term

    if nth_term%2 == 0:
      result = result + nth_term

  return result

def solution_two(limit):
  result = nth_term = 0
  first_term = second_term = 1

  while nth_term < limit:
    nth_term = first_term + second_term
    first_term = second_term + nth_term
    second_term = nth_term + first_term

    if nth_term < limit:
      result = result + nth_term

  return result

if __name__ == "__main__":
  start_time = time.time()
  #print solution_one(4000000) # Execution Time: ~[0.000024, 0.000055] seconds
  print solution_two(4000000) # Execution Time: ~[0.000019, 0.000055] seconds
  print("--- %s seconds ---" % (time.time() - start_time))